velox/dwio/parquet/reader/IAAPageReader.h

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <folly/futures/Future.h>
+
+#include "velox/dwio/common/compression/AsyncCompression.h"
+#include "velox/dwio/parquet/reader/PageReader.h"
+
+namespace facebook::velox::parquet {
+
+using folly::SemiFuture;
+
+struct PreDecompPageInfo {
+  int64_t numValues; // Number of values in this row group
+  int64_t visitedRows; // rows already read
+  const char* FOLLY_NULLABLE dataPageData{nullptr};
+  BufferPtr uncompressedData;
+};
+
+class IAAPageReader : public PageReader {
```

Comment:
Seems no need for `PageReaderBase` if you are inheriting from the concrete `PageReader`

Reply:
The reason to create ````PageReaderBase`` is to realize Polymorphism, so that when ParquetData call the same PageReader function behaves differently in different scenarios

Reply:
Reading from the code I think we can do it a little bit differently.  What you really need in `IAAPageReader` is a set of extension points (hooks) that you run extra code, in addition to the basic `PageReader`.  So you can add a few virtual methods in `PageReader`, default to no-op, call them in the expected places.  Then in `IAAPageReader` you add the implementations for these hooks.  Does it sound good to you?